vsintegration/src/FSharp.Editor/CodeLens/AbstractCodeLensDisplayService.fs

```diff
@@ -0,0 +1,273 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace rec Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.Windows.Controls
+open Microsoft.VisualStudio.Text
+open Microsoft.VisualStudio.Text.Editor
+open Microsoft.VisualStudio.Text.Formatting
+open System.Threading
+open System.Windows
+open System.Collections.Generic
+
+open Microsoft.VisualStudio.FSharp.Editor.Logging
+
+[<AbstractClass>]
+type CodeLensDisplayService (view : IWpfTextView, buffer : ITextBuffer, layerName) as self =
+    
+    // Add buffer changed event handler
+    do (
+        buffer.Changed.Add self.handleBufferChanged
+        view.LayoutChanged.Add self.handleLayoutChanged
+       )
+    
+    /// <summary>
+    /// Enqueing an unit signals to the tagger that all visible line lens must be layouted again,
+    /// to respect single line changes.
+    /// </summary>
+    member val RelayoutRequested : Queue<_> = Queue() with get
+
+    member val WpfView = view
+
+    member val TextBuffer = buffer
+
+    /// Saves the ui context to switch context for ui related work.
+    member val uiContext = SynchronizationContext.Current
+
+    // Tracks the created ui elements per TrackingSpan
+    member val uiElements = Dictionary<_,Grid>()
+
+    member val trackingSpanUiParent = HashSet()
+
+    member val uiElementNeighbour = Dictionary()
+
+    /// Caches the current used trackingSpans per line. One line can contain multiple trackingSpans
+    member val trackingSpans = Dictionary<int, ResizeArray<_>>()
+    /// Text view for accessing the adornment layer.
+    member val view: IWpfTextView = view
+    /// The code lens layer for adding and removing adornments.
+    member val codeLensLayer = view.GetAdornmentLayer layerName
+    /// Tracks the recent first + last visible line numbers for adornment layout logic.
+    member val recentFirstVsblLineNmbr = 0 with get, set
+
+    member val recentLastVsblLineNmbr = 0 with get, set
+    /// Tracks the adornments on the layer.
+    member val addedAdornments = HashSet()
+    /// Cancellation token source for the layout changed event. Needed to abort previous async-work.
+    member val layoutChangedCts = new CancellationTokenSource() with get, set
+
+    /// Tracks the last used buffer snapshot, should be preferred used in combination with mutex.
+    member val currentBufferSnapshot = null with get, set
+
+    /// Helper method which returns the start line number of a tracking span
+    member __.getTrackingSpanStartLine (snapshot:ITextSnapshot) (trackingSpan:ITrackingSpan) =
```

Comment:
Use a "let" binding for things that are essentially private methods (e.g. the ones with lowercase name)